File: uiframe/src/main/java/mobi/cangol/mobile/base/FragmentStack.java
package mobi.cangol.mobile.base;


import android.text.TextUtils;

import java.lang.ref.WeakReference;
import java.util.Stack;

/**
 * Created by xuewu.wei on 2017/10/26.
 */

public class FragmentStack {
    Stack<WeakReference<BaseFragment>> stack = null;
    Stack<String> tagStack = null;

    public FragmentStack() {
        stack = new Stack<>();
        tagStack = new Stack<>();
    }

    public int size() {
        return stack.size();
    }

    public void addFragment(BaseFragment fragment) {
        stack.add(new WeakReference<>(fragment));
    }

    public void addTag(String tag) {
        tagStack.add(tag);
    }

    public BaseFragment peekFragment() {
        return stack.isEmpty()?null:stack.peek().get();
    }

    public String peekTag() {
        return tagStack.isEmpty()?null:tagStack.peek();
    }

    public BaseFragment popFragment() {

        return stack.isEmpty()?null:stack.pop().get();
    }

    public void popFragment(String tag,int flag) {
        if(TextUtils.isEmpty(tag)){
            if(flag==1){
                while (stack.size() > 1) {
                    stack.pop();
                    tagStack.pop();
                }
            }else{
                stack.pop();
                tagStack.pop();
            }
        }else{
            int index=tagStack.indexOf(tag);
            if(index>0){
                while (!tag.equals(tagStack.peek())) {
                    tagStack.pop();
                    stack.pop();
                }
                if(flag==1){
                    tagStack.pop();
                    stack.pop();
                }
            }
        }
    }

    public String popTag() {
        return tagStack.isEmpty()?null:tagStack.pop();
    }

    public boolean containsTag(String tag){
        return tagStack.contains(tag);
    }

    public Stack<String> getTag() {
        return tagStack;
    }

    public void clear() {
        stack.clear();
        tagStack.clear();
    }
}

File: uiframe/src/main/kotlin/mobi/cangol/mobile/base/FragmentStack.kt
package mobi.cangol.mobile.base

import android.text.TextUtils
import java.util.*


/**
 * Created by xuewu.wei on 2017/10/26.
 */

class FragmentStack {
    private val stack: Stack<BaseFragment> = Stack()
    private val tag: Stack<String> = Stack()

    fun size(): Int {
        return stack.size
    }

    fun getTags(): Stack<String> {
        return tag
    }

    fun addFragment(fragment: BaseFragment) {
        stack.add(fragment)
    }

    fun addTag(tag: String) {
        this.tag.add(tag)
    }

    fun peekFragment(): BaseFragment? {
        return if (stack.isEmpty()) null else stack.peek()
    }

    fun peekTag(): String? {
        return if (tag.isEmpty()) null else tag.peek()
    }

    fun popFragment(): BaseFragment? {
        return if (stack.isEmpty()) null else stack.pop()
    }

    fun popFragment(tag: String, flag: Int) {
        if (TextUtils.isEmpty(tag)) {
            if (flag == 1) {
                while (stack.size > 1) {
                    stack.pop()
                    this.tag.pop()
                }
            } else {
                stack.pop()
                this.tag.pop()
            }
        } else {
            val index = this.tag.indexOf(tag)
            if (index > 0) {
                while (tag != this.tag.peek()) {
                    this.tag.pop()
                    stack.pop()
                }
                if (flag == 1) {
                    this.tag.pop()
                    stack.pop()
                }
            }
        }
    }

    fun popTag(): String? {
        return if (tag.isEmpty()) null else tag.pop()
    }

    fun containsTag(tag: String): Boolean {
        return this.tag.contains(tag)
    }

    fun clear() {
        stack.clear()
        tag.clear()
    }
}
